kernel/src/utilities/streaming_process_slice.rs

```diff
@@ -0,0 +1,269 @@
+///    then issues an `allow` operation that atomically swaps the current
+///    allowed buffer by its second buffer.
+///
+/// 7. The process can now process the received chunks contained in the initial
+///    buffer, while the kernel receives new chunks in the other, newly allowed
+///    buffer.
+///
+///
+/// The version 0 buffer format is specified as follows:
+/// ```text,ignore
+/// 0          1          2          3          4          5
+/// +----------+----------+----------+----------+----------+-------------------...
+/// | flags    | buffer offset                             | data
+/// +----------+----------+----------+----------+----------+-------------------...
+/// | VVVVxxHE | 32 bits (native endian)                   |
+/// ```
+///
+/// The flags field is structured as follows:
+/// - `V`: version bits. This kernel only supports version `0`.
+/// - `H`: `halt` flag. If this flag is set alongside the `exceeded` flag, the
+///   kernel will not write any further data to this buffer.
+/// - `E`: `exceeded` flag. The kernel sets this flag when the remaining buffer
+///   capacity is insufficient to append the current chunk.
+/// - `x`: reserved flags. Unless specified otherwise, processes must clear
+///   these flags prior to `allow`ing a buffer to the kernel. A kernel that does
+///   not know of a reserved flag must refuse to operate on a buffer that has
+///   such a flag set.
+#[repr(transparent)]
+pub struct StreamingProcessSlice<'a> {
+    slice: &'a WriteableProcessSlice,
+}
+
+register_bitfields![
+    u8,
+    pub StreamingProcessSliceFlags [
+        VERSION OFFSET(4) NUMBITS(4) [
+            V0 = 0x00,
+        ],
+        RESERVED OFFSET(2) NUMBITS(2) [
+            RESERVED0 = 0x00,
+        ],
+        HALT OFFSET(1) NUMBITS(1) [],
+        EXCEEDED OFFSET(1) NUMBITS(1) [],
+    ]
+];
+
+#[derive(Debug, Clone, Copy)]
+pub struct StreamingProcessSliceMeta {
+    pub version: StreamingProcessSliceFlags::VERSION::Value,
+    pub halt: bool,
+    pub exceeded: bool,
+    pub offset: u32,
+}
+
+impl<'a> StreamingProcessSlice<'a> {
+    const RANGE_FLAGS: Range<usize> = 0..1;
+    const RANGE_OFFSET: Range<usize> = (Self::RANGE_FLAGS.end)..(Self::RANGE_FLAGS.end + 4);
+    const RANGE_DATA: RangeFrom<usize> = (Self::RANGE_OFFSET.end + 1)..;
+
+    pub fn new(slice: &'a WriteableProcessSlice) -> StreamingProcessSlice {
+        StreamingProcessSlice { slice }
+    }
+
+    /// Checks whether the buffer is valid (of sufficient size to contain at
+    /// least the `flags` and `offset` fields), and extract the `flags` and
+    /// `offset` field values.
+    ///
+    /// This function fails with
+    /// - `INVAL`: if the version is not `0`, or the reserved flags are not
+    ///   cleared.
+    /// - `SIZE`: if the underlying slice is not large enough to fit the
+    ///   flags field and the offset field (5 bytes).
+    pub fn get_meta(&self) -> Result<StreamingProcessSliceMeta, ErrorCode> {
```

Comment:
Why is this and `set_meta()` public? This exposes the version semantics to capsules which if we ever wanted a version 1 could mean we would have to edit all capsules that use this abstraction.

Maybe we do need these to be public, but if we are not using them right now I think it safer to not expose these two APIs.

Reply:
Yes, I think those shouldn't be. I agree that we should only provide the higher-level interfaces, and add more if required that abstract over the version & flag details.